src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4451,6 +4451,76 @@ public override IOperation VisitMemberInitializer(IMemberInitializerOperation op
             throw ExceptionUtilities.Unreachable;
         }
 
+        public override IOperation VisitArrayCreation(IArrayCreationOperation operation, int? captureIdForResult)
+        {
+            // We have couple of options on how to rewrite an array creation with an initializer:
+            //       1) Retain the original tree shape so the visited IArrayCreationOperation still has an IArrayInitializerOperation child node.
+            //       2) Lower the IArrayCreationOperation so it always has a null initializer, followed by explicit assignments
+            //          of the form "IArrayElementReference = value" for the array initializer values.
+            //          There will be no IArrayInitializerOperation in the tree with approach.
+            //
+            //  We are going ahead with approach #1 for couple of reasons:
+            //  1. Simplicity: The implementation is much simpler, and has a lot lower risk associated with it.
+            //  2. Lack of array instance access in the initializer: Unlike the object/collection initializer scenario,
+            //     where the initializer can access the instance being initialized, array initializer does not have access
+            //     to the array instance being initialized, and hence it does not matter if the array allocation is done
+            //     before visiting the initializers or not.
+            //
+            //  In future, based on the customer feedback, we can consider switching to approach #2 and lower the initializer into assignment(s).
+
+            PushArray(operation.DimensionSizes);
+            IArrayInitializerOperation visitedInitializer = Visit(operation.Initializer);
+            ImmutableArray<IOperation> visitedDimensions = PopArray(operation.DimensionSizes);
+            return new ArrayCreationExpression(visitedDimensions, visitedInitializer, semanticModel: null,
+                                               operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitArrayInitializer(IArrayInitializerOperation operation, int? captureIdForResult)
+        {
+            VisitAndPushArrayInitializerValues(operation);
+            return PopAndAssembleArrayInitializerValues(operation);
+        }
+
+        private void VisitAndPushArrayInitializerValues(IArrayInitializerOperation operation)
```

Comment:
Consider making these local functions, unless you think there's a compelling reason to reuse outside VisitArrayInitializer.

Reply:
Personally, I don't link using local functions when they are recursive and one of the parameter types matches the outer function's type. I am fine to switch if compiler team prefers local function for such cases.

Reply:
My 2 cents: in bug files like this where there are lots of necessary member functions I prefer to use local functions whenever possible. This keeps the member list as small as possible which I think helps with understanding the impact of changes. 

---
In reply to: [188758281](https://github.com/dotnet/roslyn/pull/26901#discussion_r188758281) [](ancestors = 188758281)

Reply:
Sure, I can make these local functions with my next commit/PR.

PS: I wish the language allowed me to explicitly state _I do not want this local function to have access to symbols  from the containing function (locals and parameters)_, so I can prevent accidental usage of containing function parameter instead of local function parameter, which I end up doing very often.

Reply:
I want the same feature. Please email LDM to express this feedback. Every response helps ;) 

---
In reply to: [188764605](https://github.com/dotnet/roslyn/pull/26901#discussion_r188764605) [](ancestors = 188764605)